File: python/src/contriwork_market_data/adapters/iex_cloud.py
"""IEX Cloud adapter — REST surface against the v1 API.

The original IEX Cloud service was sunset in August 2024; the surface
shape is preserved here so callers running compatible providers (e.g.
mirror deployments or replacement vendors who kept the path scheme) can
swap base_url to redirect.

Endpoints used:
- ``/stable/stock/{symbol}/quote`` — spot.
- ``/stable/stock/{symbol}/chart/{range}`` — historical candles.

Auth header: ``token`` query parameter or ``Authorization`` header.
"""

from __future__ import annotations

from collections.abc import AsyncIterator, Awaitable, Callable
from datetime import UTC, datetime
from decimal import Decimal
from typing import Any

import httpx

from .._adapter_helpers import resolve_api_key, streaming_not_supported
from .._http import build_async_client, request_json
from ..errors import (
    AdapterFeatureNotSupportedError,
    AdapterUnavailableError,
    InvalidIntervalError,
    SymbolNotFoundError,
)
from ..types import (
    Candle,
    Capability,
    Interval,
    OrderBook,
    SpotPrice,
    Ticker,
)

__all__ = ["IEXCloudAdapter"]

_DEFAULT_BASE = "https://cloud.iexapis.com"
_RATE_LIMIT_PER_MINUTE = 100

_RANGE_FOR_INTERVAL: dict[Interval, str] = {
    Interval.D1: "1m",
    Interval.W1: "1y",
    Interval.MN1: "max",
}

_SUPPORTED_INTERVALS: tuple[Interval, ...] = tuple(_RANGE_FOR_INTERVAL.keys())


class IEXCloudAdapter:
    adapter_id = "iex-cloud"

    def __init__(
        self,
        *,
        api_key: str | None = None,
        api_key_provider: Callable[[], Awaitable[str | None]] | None = None,
        base_url: str | None = None,
        timeout_s: float = 15.0,
        http_proxy: str | None = None,
        http_client: httpx.AsyncClient | None = None,
    ) -> None:
        self._api_key = api_key
        self._api_key_provider = api_key_provider
        self._base_url = (base_url or _DEFAULT_BASE).rstrip("/")
        self._timeout_s = timeout_s
        self._http_proxy = http_proxy
        self._client = http_client
        self._owns_client = http_client is None
        self.capability = Capability(
            supported_markets=("stocks_us",),
            supported_intervals=_SUPPORTED_INTERVALS,
            supported_quote_currencies=("USD",),
            supports_order_book=False,
            supports_native_streaming=False,
            rate_limit_per_minute=_RATE_LIMIT_PER_MINUTE,
            requires_auth=True,
            tier_options=("sandbox", "standard"),
        )

    async def aclose(self) -> None:
        if self._owns_client and self._client is not None:
            await self._client.aclose()
            self._client = None

    async def _ensure_client(self) -> httpx.AsyncClient:
        if self._client is None:
            self._client = build_async_client(
                timeout_s=self._timeout_s,
                http_proxy=self._http_proxy,
            )
        return self._client

    async def _api_key_value(self) -> str:
        key = await resolve_api_key(
            adapter_id=self.adapter_id,
            api_key=self._api_key,
            api_key_provider=self._api_key_provider,
            required=True,
        )
        assert key is not None
        return key

    async def get_spot(self, symbol: str, quote_currency: str) -> SpotPrice:
        client = await self._ensure_client()
        key = await self._api_key_value()
        try:
            payload = await request_json(
                client,
                "GET",
                f"{self._base_url}/stable/stock/{symbol}/quote",
                adapter_id=self.adapter_id,
                params={"token": key},
            )
        except AdapterUnavailableError as exc:
            if "HTTP 404" in str(exc):
                raise SymbolNotFoundError(
                    f"iex-cloud does not know symbol {symbol!r}",
                    adapter_id=self.adapter_id,
                ) from exc
            raise
        if not isinstance(payload, dict) or "latestPrice" not in payload:
            raise AdapterUnavailableError(
                "iex-cloud returned unexpected payload",
                adapter_id=self.adapter_id,
            )
        return SpotPrice(
            symbol=symbol,
            last=Decimal(str(payload["latestPrice"])),
            quote_currency=quote_currency,
            timestamp=datetime.fromtimestamp(int(payload.get("latestUpdate", 0)) / 1000, tz=UTC),
            source_adapter=self.adapter_id,
            bid=_opt_decimal(payload.get("iexBidPrice") or payload.get("bidPrice")),
            ask=_opt_decimal(payload.get("iexAskPrice") or payload.get("askPrice")),
            high_24h=_opt_decimal(payload.get("high")),
            low_24h=_opt_decimal(payload.get("low")),
            volume_24h=_opt_decimal(payload.get("latestVolume")),
            change_24h_pct=_opt_decimal(payload.get("changePercent")),
            previous_close=_opt_decimal(payload.get("previousClose")),
            market_cap=_opt_decimal(payload.get("marketCap")),
        )

    async def get_ohlcv(
        self,
        symbol: str,
        interval: Interval,
        since: datetime | None,
        limit: int,
    ) -> list[Candle]:
        if interval not in _SUPPORTED_INTERVALS:
            raise InvalidIntervalError(
                f"iex-cloud does not support interval {interval.value}",
                adapter_id=self.adapter_id,
            )
        client = await self._ensure_client()
        key = await self._api_key_value()
        range_value = _RANGE_FOR_INTERVAL[interval]
        try:
            payload = await request_json(
                client,
                "GET",
                f"{self._base_url}/stable/stock/{symbol}/chart/{range_value}",
                adapter_id=self.adapter_id,
                params={"token": key, "chartCloseOnly": "false"},
            )
        except AdapterUnavailableError as exc:
            if "HTTP 404" in str(exc):
                raise SymbolNotFoundError(
                    f"iex-cloud does not know symbol {symbol!r}",
                    adapter_id=self.adapter_id,
                ) from exc
            raise
        if not isinstance(payload, list):
            raise AdapterUnavailableError(
                "iex-cloud chart returned unexpected payload",
                adapter_id=self.adapter_id,
            )
        candles: list[Candle] = []
        for row in payload[:limit]:
            ts = datetime.strptime(row["date"], "%Y-%m-%d").replace(tzinfo=UTC)
            if since is not None and ts < since:
                continue
            candles.append(
                Candle(
                    timestamp=ts,
                    open=Decimal(str(row["open"])),
                    high=Decimal(str(row["high"])),
                    low=Decimal(str(row["low"])),
                    close=Decimal(str(row["close"])),
                    volume=Decimal(str(row.get("volume", 0))),
                )
            )
        return candles

    async def get_order_book(self, symbol: str, depth: int) -> OrderBook:
        raise AdapterFeatureNotSupportedError(
            "iex-cloud does not expose a public order book endpoint",
            adapter_id=self.adapter_id,
        )

    def subscribe_ticker(self, symbol: str) -> AsyncIterator[Ticker]:
        return streaming_not_supported(self.adapter_id)


def _opt_decimal(value: Any) -> Decimal | None:
    if value is None or value == "":
        return None
    return Decimal(str(value))

File: typescript/src/adapters/iex-cloud.ts
/**
 * IEX Cloud adapter — preserves the v1 path scheme (the service was sunset
 * in August 2024; compatible mirrors can swap baseUrl). `/stable/stock/{s}/quote`
 * + `/chart/{range}`.
 */
import { Decimal } from "decimal.js";
import type { ApiKeyProvider, MarketDataAdapter } from "../adapter.js";
import { resolveApiKey } from "../internal/credentials.js";
import {
  type FetchLike,
  defaultFetch,
  getJson,
  queryString,
} from "../internal/http.js";
import { streamingNotSupported } from "../internal/streaming.js";
import {
  AdapterFeatureNotSupportedError,
  AdapterUnavailableError,
  InvalidIntervalError,
  SymbolNotFoundError,
} from "../errors.js";
import {
  type Candle,
  type Capability,
  EMPTY_EXTRA,
  Interval,
  type OrderBook,
  type SpotPrice,
  type Ticker,
} from "../types.js";

const RANGE_FOR_INTERVAL = new Map<Interval, string>([
  [Interval.D1, "1m"],
  [Interval.W1, "1y"],
  [Interval.MN1, "max"],
]);

/** Options for {@link IEXCloudAdapter}. */
export interface IEXCloudOptions {
  readonly apiKey?: string;
  readonly apiKeyProvider?: ApiKeyProvider;
  readonly baseUrl?: string;
  readonly fetchFn?: FetchLike;
}

function optDv(value: unknown): Decimal | undefined {
  return value === undefined || value === null
    ? undefined
    : new Decimal(value as Decimal.Value);
}

/** IEX Cloud US-stocks adapter. */
export class IEXCloudAdapter implements MarketDataAdapter {
  public readonly adapterId = "iex-cloud";
  public readonly capability: Capability;

  private readonly apiKey?: string;
  private readonly apiKeyProvider?: ApiKeyProvider;
  private readonly baseUrl: string;
  private readonly fetchFn: FetchLike;

  public constructor(options: IEXCloudOptions = {}) {
    if (options.apiKey !== undefined) {
      this.apiKey = options.apiKey;
    }
    if (options.apiKeyProvider !== undefined) {
      this.apiKeyProvider = options.apiKeyProvider;
    }
    this.baseUrl = (options.baseUrl ?? "https://cloud.iexapis.com").replace(
      /\/+$/,
      "",
    );
    this.fetchFn = options.fetchFn ?? defaultFetch;
    this.capability = {
      supportedMarkets: ["stocks_us"],
      supportedIntervals: [...RANGE_FOR_INTERVAL.keys()],
      supportedQuoteCurrencies: ["USD"],
      supportsOrderBook: false,
      supportsNativeStreaming: false,
      rateLimitPerMinute: 100,
      requiresAuth: true,
      tierOptions: ["sandbox", "standard"],
    };
  }

  private async key(signal?: AbortSignal): Promise<string> {
    return (await resolveApiKey(
      this.adapterId,
      this.apiKey,
      this.apiKeyProvider,
      true,
      signal,
    )) as string;
  }

  private async get(
    path: string,
    query: Record<string, string | number | undefined>,
    signal?: AbortSignal,
  ): Promise<unknown> {
    try {
      return await getJson(
        this.fetchFn,
        this.adapterId,
        this.baseUrl + path + queryString(query),
        undefined,
        signal,
      );
    } catch (err) {
      if (
        err instanceof AdapterUnavailableError &&
        err.message.includes("HTTP 404")
      ) {
        throw new SymbolNotFoundError(
          `iex-cloud does not know symbol (404): ${path}`,
          this.adapterId,
        );
      }
      throw err;
    }
  }

  public async getSpot(
    symbol: string,
    quoteCurrency: string,
    signal?: AbortSignal,
  ): Promise<SpotPrice> {
    const key = await this.key(signal);
    const root = (await this.get(
      `/stable/stock/${symbol}/quote`,
      { token: key },
      signal,
    )) as Record<string, unknown>;
    if (root["latestPrice"] === undefined) {
      throw new AdapterUnavailableError(
        "iex-cloud returned unexpected payload",
        this.adapterId,
      );
    }
    const bid = optDv(root["iexBidPrice"] ?? root["bidPrice"]);
    const ask = optDv(root["iexAskPrice"] ?? root["askPrice"]);
    const high = optDv(root["high"]);
    const low = optDv(root["low"]);
    const vol = optDv(root["latestVolume"]);
    const chg = optDv(root["changePercent"]);
    const prev = optDv(root["previousClose"]);
    const mcap = optDv(root["marketCap"]);
    return {
      symbol,
      last: new Decimal(root["latestPrice"] as Decimal.Value),
      quoteCurrency,
      timestamp: new Date(Number(root["latestUpdate"] ?? 0)),
      sourceAdapter: this.adapterId,
      extra: EMPTY_EXTRA,
      ...(bid !== undefined && { bid }),
      ...(ask !== undefined && { ask }),
      ...(high !== undefined && { high24h: high }),
      ...(low !== undefined && { low24h: low }),
      ...(vol !== undefined && { volume24h: vol }),
      ...(chg !== undefined && { change24hPct: chg }),
      ...(prev !== undefined && { previousClose: prev }),
      ...(mcap !== undefined && { marketCap: mcap }),
    };
  }

  public async getOhlcv(
    symbol: string,
    interval: Interval,
    since: Date | undefined,
    limit: number,
    signal?: AbortSignal,
  ): Promise<readonly Candle[]> {
    const range = RANGE_FOR_INTERVAL.get(interval);
    if (range === undefined) {
      throw new InvalidIntervalError(
        `iex-cloud does not support interval ${interval}`,
        this.adapterId,
      );
    }
    const key = await this.key(signal);
    const payload = await this.get(
      `/stable/stock/${symbol}/chart/${range}`,
      { token: key, chartCloseOnly: "false" },
      signal,
    );
    if (!Array.isArray(payload)) {
      throw new AdapterUnavailableError(
        "iex-cloud chart returned unexpected payload",
        this.adapterId,
      );
    }
    const candles: Candle[] = [];
    for (const row of payload as Record<string, unknown>[]) {
      const ts = new Date(`${String(row["date"])}T00:00:00Z`);
      if (since !== undefined && ts.getTime() < since.getTime()) {
        continue;
      }
      candles.push({
        timestamp: ts,
        open: new Decimal(row["open"] as Decimal.Value),
        high: new Decimal(row["high"] as Decimal.Value),
        low: new Decimal(row["low"] as Decimal.Value),
        close: new Decimal(row["close"] as Decimal.Value),
        volume:
          row["volume"] !== undefined
            ? new Decimal(row["volume"] as Decimal.Value)
            : new Decimal(0),
        extra: EMPTY_EXTRA,
      });
      if (candles.length >= limit) {
        break;
      }
    }
    return candles;
  }

  public getOrderBook(
    _symbol: string,
    _depth: number,
    _signal?: AbortSignal,
  ): Promise<OrderBook> {
    return Promise.reject(
      new AdapterFeatureNotSupportedError(
        "iex-cloud does not expose a public order book endpoint",
        this.adapterId,
      ),
    );
  }

  public subscribeTicker(
    _symbol: string,
    _signal?: AbortSignal,
  ): AsyncIterable<Ticker> {
    return streamingNotSupported(this.adapterId);
  }
}
